crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -0,0 +1,206 @@
+                .with_url(destination.clone())
+                .with_region(aws_region.to_string())
+                .with_credentials(Arc::new(AwsSdkCredentialsProvider {
+                    credentials_provider: DefaultCredentialsChain::builder().build().await,
+                }))
+                .build()?;
+
+            Arc::new(store)
+        } else {
+            object_store::parse_url_opts(&destination, &snapshots_options.additional_options)?
+                .0
+                .into()
+        };
+
+        let prefix = destination.path().into();
+        Ok(SnapshotRepository {
+            object_store,
+            destination,
+            prefix,
+            staging_path: base_dir.clone().join("snapshot-staging"),
+        })
+    }
+
+    /// Write a partition snapshot to the snapshot repository.
+    pub(crate) async fn put(
+        &self,
+        partition_id: PartitionId,
+        metadata: &PartitionSnapshotMetadata,
+        snapshot_path: &Path,
+    ) -> anyhow::Result<()> {
+        let snapshot_id = metadata.snapshot_id;
+        let lsn = metadata.min_applied_lsn;
+
+        debug!(
+            %snapshot_id,
+            partition_id = ?partition_id,
+            %lsn,
+            "Publishing partition snapshot to: {}",
+            self.destination,
+        );
+
+        // All common object stores list objects in lexicographical order, with no option for
+        // reverse order. We inject an explicit sort key into the snapshot prefix to make sure that
+        // the latest snapshot is always first.
+        let inverted_sort_key = format!("{:016x}", u64::MAX - lsn.as_u64());
+
+        // The snapshot data / metadata key format is: [<base_prefix>/]<partition_id>/<sort_key>_<lsn>_<snapshot_id>.tar
+        let snapshot_key = match self.prefix.as_str() {
+            "" | "/" => format!(
+                "{partition_id}/{sk}_{lsn}_{snapshot_id}.tar",
+                sk = inverted_sort_key,
+                lsn = metadata.min_applied_lsn,
+            ),
+            prefix => format!(
+                "{trimmed_prefix}/{partition_id}/{sk}_{lsn}_{snapshot_id}.tar",
+                trimmed_prefix = prefix.trim_start_matches('/').trim_end_matches('/'),
+                sk = inverted_sort_key,
+            ),
+        };
+
+        let staging_path = self.staging_path.clone();
+        tokio::fs::create_dir_all(&staging_path).await?;
+
+        let snapshot_path = snapshot_path.to_path_buf();
+        let packaging_task = task_center().spawn_blocking_unmanaged(
+            "package-snapshot",
+            Some(partition_id),
+            async move {
+                trace_span!("package-snapshot", %snapshot_id).in_scope(|| {
+                    let mut tarball = tar::Builder::new(NamedTempFile::new_in(&staging_path)?);
+                    debug!(
+                        "Creating snapshot tarball of {:?} in: {:?}...",
+                        &staging_path,
```

Comment:
```suggestion
                        &snapshot_path,
```

?

Reply:
I've renamed `staging_path` to `local_snapshot_path` for clarity - that's the raw RocksDB column family export directory with the SSTs plus our own metadata JSON blob. We then tar that directory up into an archive at the path `snapshot_archive_path`.